File: 19.Backend Scaling and Performance Engineering - Part-2/code/typescript/example_3.ts
import Redis from 'ioredis';

const rdb = new Redis();

interface EmailJob {
    to: string;
    subject: string;
    template: string;
}

export async function enqueueEmail(job: EmailJob) {
    await rdb.lpush("queue:emails", JSON.stringify(job));
}

export async function emailWorker() {
    while (true) {
        const result = await rdb.brpop("queue:emails", 0);
        if (result) {
            const job: EmailJob = JSON.parse(result[1]);
            // sendEmail(job) takes 300ms, but no user is waiting
        }
    }
}

File: 19.Backend Scaling and Performance Engineering - Part-2/code/rust/example_3.rs
use redis::AsyncCommands;
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize)]
pub struct EmailJob {
    pub to: String,
    pub subject: String,
    pub template: String,
}

pub async fn enqueue_email(mut con: redis::aio::Connection, job: EmailJob) -> redis::RedisResult<()> {
    let data = serde_json::to_string(&job).unwrap();
    con.lpush("queue:emails", data).await
}

pub async fn email_worker(mut con: redis::aio::Connection) {
    loop {
        let result: redis::RedisResult<Vec<String>> = con.brpop("queue:emails", 0.0).await;
        if let Ok(res) = result {
            if res.len() == 2 {
                let job: EmailJob = serde_json::from_str(&res[1]).unwrap();
                // send_email(job) takes 300ms, but no user is waiting
            }
        }
    }
}
